integreat_cms/summ_ai_api/utils.py

```diff
@@ -392,3 +394,103 @@ def __repr__(self):
         :rtype: str
         """
         return f"<TranslationHelper (translation: {self.german_translation!r})>"
+
+
+class PatientTaskQueue(deque):
+    """
+    A 'patient' task queue which only hands out sleep tasks after a task was reported as failed.
+
+    @TODO: sphinx throws warnings (PatientTaskQueue.count:1: WARNING: py:class reference target not found:
+           integer -- return number of occurrences of value), because it tries to document the deque superclass with.
+           Neither
+           - .. autodoc_exclude_members: count, index
+           - .. autodoc_exclude_members: deque
+           - .. autodoc_exclude_class: deque
+           - .. autodoc_exclude_class: PatientTaskQueue
+           solved that problem. Any suggestions?
+    """
+
+    #: The UNIX timestamp when the last rate limited request occured
+    #:
+    #: :type: int
+    last_rate_limit = None
+    #: The UNIX timestamp when the last rate limited request occured
+    #:
+    #: :type: int
+    wait_time = settings.SUMM_AI_WAIT_FOR_NEXT_REQUEST_AFTER_LIMIT_EXCEEDED
+
+    def __init__(
+        self,
+        tasks,
+        wait_time=settings.SUMM_AI_WAIT_FOR_NEXT_REQUEST_AFTER_LIMIT_EXCEEDED,
+    ):
+        """
+        Constructor initializes the class variables
+
+        :param tasks: List of request tasks
+        :type tasks: list [asyncio.Future]
+
+        :param wait_time: Waiting time until start next request in seconds
+        :type wait_time: int
+        """
+        super().__init__(tasks)
+        self.last_rate_limit = None
+        self.wait_time = wait_time
+
+    def __next__(self):
+        """
+        Extension of the next() function of deque
+        Checks if the queue processing should wait.
+        Ejects the next task or goes to sleep until the end of the waiting time.
+
+        :returns: a task of the queue
+        :rtype: asyncio.Future
+        """
+        now = time.time()
+        if (
+            self.last_rate_limit is not None
+            and now - self.last_rate_limit < self.wait_time
+        ):
+            # If we are currently waiting out a rate limit,
+            # the worker should sleep for the remaining time before getting the next real task
+            return asyncio.sleep(self.wait_time - (now - self.last_rate_limit))
```

Comment:
Since we're not sure of the larger scale implications of this and decide we are not going to spend time on testing alternative solutions, we will not handle this for now.

Reply:
Alright, fine for me

Reply:
Thought about this a bit more after all. Since the returned text fields are never used (as they are the same objects that were initially given to the `PatientTaskQueue`) it doesn't really matter when there are a few extra `None` among them, and if they are going to be, they will be easy to filter out – but yes, it does feel hacky. I'm flinching away from letting the whole thread sleep for this long of a time span (without having a concrete understanding of the consequences) though, and experimented with making `PatientTaskQueue` an asynchronous iterator as well.